<a href="https://colab.research.google.com/github/bzznrc/Python-101/blob/master/FetchDemo.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Fetch Demo 🐕
Oook. Bit of background needed.

We're now working with some real-world data from one of our real-world partners. The reason we're importing this data with a very old fashioned .csv is because these data are quite sensitive, so we couldn't really dump them into the GitHub.

The bits of code that follow just let you upload a file from you own PC and load its content into a DataFrame. You don't really need to understand how they work.

In [1]:
from google.colab import files
import pandas as pd
import numpy as np
import io
#Magical bit of code you shouldn't really worry about
uploaded = files.upload()
df = pd.read_csv(io.StringIO(uploaded['FetchUS_BAR_V2.csv'].decode('utf-8')))

Saving FetchUS_BAR_V2.csv to FetchUS_BAR_V2.csv


#But What is a DataFrame? 🖼
Right. So you can see there were a couple of libraries imported there. Pandas is the most important out of them, and it's by far the #1 used library for all the Data Scientists in the world.
The main thing it does, it's introducing this DataFrame concept into Python, and all the functionalities to work with it.

A DataFrame is, basically, a data table. And you can think of it as an "Excel within Python", if you'd like. As you can see in the output above, the column names are the ones coming from your .csv while the rows are indexed with a number (the one on the left).

Our DataFrame is called "df", as pretty much every other DataFrame in every notebook in the world. We're Data Scientists, so you can guess creativity is not top-of-mind.

In [11]:
#Let's see how big our DataFrame is using the shape attribute. Be aware, this is not a method, so there are no parenthesis required
#An attribute is basically a label that your object (your df in this case) always carries
print("DF Shape (rows, columns): ", df.shape)

#If you need either the number of rows or columns on their own:
#df.shape[0] #Gets you the nr. of rows
#df.shape[1] #Gets you the nr. of columns

DF Shape (rows, columns):  (92667, 9)


In [12]:
#When you have a DataFrame the first thing you want to do is see what's inside.
#The head() method shows you the first 5 lines
df.head()

store_name            store_city store_state                   user_id  \
0  SAM'S CLUB          Calumet City          IL  600d813fadf77112c3bfa9a1   
1     SMITH'S                  Lehi          UT  60087e70c658dc12dc7354e1   
2     WALMART       West Palm Beach          FL  5fa6dbdcf54a65129aff01e9   
3         Y U  Feasterville Trevose          PA  5fbbe808fc5c9221743b89ce   
4      KROGER             Nashville          TN  5f38942e73dd9a14dbf9d914   

   transaction_date  total_sales  total_qty  barilla_sales  barilla_qty  
0  09/09/2021 11:04        79.51         12           6.54          1.0  
1  07/09/2021 16:08       151.97         51           1.59          1.0  
2  11/09/2021 09:52       257.16         86           1.28          1.0  
3  13/09/2021 11:38         1.89          1           1.89          1.0  
4  15/09/2021 18:07       274.64         86           1.59          1.0

In [13]:
#Unimpressingly, the tail() method shows you the last 5 lines
df.tail()

store_name    store_city store_state  \
92662                 TARGET   Port Orange          FL   
92663  VALLARTA SUPERMARKETS        Fresno          CA   
92664              FOOD MAXX        Fresno          CA   
92665                 PUBLIX  Coral Gables          FL   
92666       THE FRESH GROCER           NaN         NaN   

                        user_id  transaction_date  total_sales  total_qty  \
92662  600ae6c7851f1c12afad3096  19/09/2021 19:29        59.36         19   
92663  5fcaf117e3824d12a17333b5  17/09/2021 14:49        94.22         41   
92664  5fcaf117e3824d12a17333b5  23/09/2021 21:14        27.80         15   
92665  5cd22a4695c514319b4b15b3  28/09/2021 00:00       165.45         32   
92666  5f78de338acf3d16579dcc3a  24/09/2021 14:23       219.87         54   

       barilla_sales  barilla_qty  
92662           1.29          1.0  
92663           1.89          1.0  
92664           1.29          1.0  
92665           3.70          2.0  
92666          22.67          2.0

In [9]:
#The describe method provide you with a summary of how the amounts are distributed in the numerical columns
#This is very useful to quickly assess the range of values, what the averages are, how narrow the distributions might be
df.describe()

total_sales     total_qty  barilla_sales   barilla_qty
count   92667.000000  92667.000000   92627.000000  92663.000000
mean      142.443297     34.929036       3.126895      1.592966
std      1990.782310     31.369580      15.780766      1.180888
min         0.000000      1.000000       0.000000      0.000000
25%        48.700000     16.000000       1.400000      1.000000
50%        93.110000     29.000000       2.190000      1.000000
75%       159.010000     47.000000       3.580000      2.000000
max    310165.190000   3085.000000    4313.580000     69.000000

#Selecting Stuff 👉
So you have a bunch of data into a DataFrame. What your inner Data Analyst will likely want to do is either selecting and displaying data that match a certain criteria and then plotting the results.

Let's start with selecting stuff. Let's say we want only the Cracker products.

This kinda looks like what we did with if conditions, but here when you write a condition, it is applied to every row. This is how you ask Pandas whether each line is a Store in San Francisco or not:

In [40]:
#Weeell... there are some San Francisco rows somewhere, believe me
df['store_city'] == 'San Francisco'

#Ok ok, so "believe me" is not want you want to hear from your Data Scientist... here is how you make sure:
#print((df['store_city'] == 'San Francisco').value_counts())

0        False
1        False
2        False
3        False
4        False
         ...  
92662    False
92663    False
92664    False
92665    False
92666    False
Name: store_city, Length: 92667, dtype: bool

Cool! So now we have 92666 True or False values and we don't have a clue what to do with them!

Kidding, now we'll store the information that we got in a Series (which is, a DataFrame with a single column) and use it to select actual rows for San Francisco stores in the main DataFrame.

Basically what happens is that the DataFrame will only show rows that correspond to the "True" values in the Series:

In [18]:
#Same code as before, getting the True/False for our condition for all rows
isSanFrancisco = df['store_city'] == 'San Francisco'

#Creating a new DataFrame with the filter applied
df_SF = df[isSanFrancisco]
print("Number of stores in San Francisco: ", df_SF.shape[0]) #Getting the first element of the (rows, columns) pair

#Showing the resulting DF
df_SF.head()

Number of stores in San Francisco:  43


store_name     store_city store_state                   user_id  \
3347         SAFEWAY  San Francisco          CA  5f1cf6f69a6b7314bd04c826   
4105         SAFEWAY  San Francisco          CA  5fed3d1d21684312583e1318   
4211        FOODS CO  San Francisco          CA  5faa149325538012adefe373   
4540         SAFEWAY  San Francisco          CA  5fd55f352909db4c05e40f04   
6922  MOLLIE STONE'S  San Francisco          CA  6013059d49a28312fd9a8db0   

      transaction_date  total_sales  total_qty  barilla_sales  barilla_qty  
3347  12/01/2022 00:00       322.39         49           2.49          1.0  
4105  15/01/2022 15:06        14.64          6           2.49          1.0  
4211  13/01/2022 18:20       294.00         96           7.96          4.0  
4540  06/01/2022 09:51        14.38          7          10.95          5.0  
6922  06/01/2022 14:15       292.87         57           7.77          3.0

# Grouping Stuff 🥂
How nice it is when our data come exactly in the format we need them into?
Not sure, never happened to me. 🤷‍♂️

Most likely, when you have a DataFrame you'll want to produce a few additional ones with different kinds of grouping, in order to see suff from different perspectives.

For example, with starting from the original DataFrame we could ask:
- How many different Stores do we have in each City?
- How much total_sales every Store has?

In order to answer to both questions, we'll definitely need to **group** the data (in the first example by *store_city* and in the second one by *store_name*).

Grouping the data by a field (or more than one fields) means producing a new DataFrame in which each row will correspond to only one of the grouping values.

Let's see an example:

In [24]:
#First, we group the df by 'store_city'
#Then we consider only the 'store_name' column
#On that column, we count how many occurrences we have (for each group)
df.groupby(by='store_city')['store_name'].count()

#Mhhh... There seem to be some dirty store_city values. We can sort the results showing the biggest ones first:
#df.groupby(by='store_city')['store_name'].count().sort_values(ascending=False)

store_city
Orlando        622
Houston        596
San Antonio    516
Miami          443
Chicago        405
              ... 
LOS BANOS        1
LORIS            1
LOOGOOTEE        1
LONGWOOD         1
|RENSSELAER      1
Name: store_name, Length: 10166, dtype: int64

In [43]:
#First, we group the df by 'category_3'
#Then we consider only the 'description' column
#On that column, we sum up the total_sales (for each group)
df.groupby(by='store_name')['total_sales'].sum()

#Mhhh... same thing here. Let's sort:
#df.groupby(by='store_name')['total_sales'].sum().sort_values(ascending=False)

store_name
00LLAR GENERAL                    48.68
06LICPOA BRDRSTER MARKET          96.08
1 6 FEDE AL RUAI                 345.61
1 DERMATOLOGY SPECIALISTS, PA    191.25
1 OZ COFFEE, LLC                   5.03
                                  ...  
smart & final                     57.84
smart and final                  310.80
stop. and shop                    45.76
super king                        26.14
weis                              59.75
Name: total_sales, Length: 1703, dtype: float64

### Plotting 🖌🎨
Sooo tables are nice, but let's be real. You have to be able to build fancy plots to be taken seriously.

But don't worry, there are libraries that basically do the work for you. Just a few names:
- Matplotlib
- Plotly
- Seaborn

In the context of this excercise, we'll use Plotly Express, which is probably the simplest out there.

Let's use it to plot the last one of the previous excercise outputs. Try to uncomment the three versions of the plot in order to appreciate the improvements.

In [45]:
#First we import the library like this:
import plotly.express as px

#This is the result from the previous excercise. Let's save it in a DataFrame
df_store_sales = df.groupby(by='store_name')['total_sales'].sum()

# 1 - Then, we'll use the histogram function to plot it right away
#px.bar(df_store_sales)

# 2 - Nope. Doesn't work. Let's try to sort the values and just keep the biggest ones
#px.bar(df_store_sales.sort_values(ascending=False)[0:25])

# 3 - That's better. Let's try to make a few cosmetic changes
#The labels parameter lets you use a Dictionary to rename the axis. The update_layout method lets you remove the legend
px.bar(df_store_sales.sort_values(ascending=False)[0:25], labels={'value':'Total Sales', 'store_name':'Store Name'}).update_layout(showlegend=False)

# Real-World Example
Let's do a basic real-world analysis:
- Build a custom column for the Barilla market share
- Select the top 25 stores in California
- For those 25 stores, plot the Barilla market share, sorted

In [46]:
#Recap: this is the DataFrame we start with:
df.head()

store_name            store_city store_state                   user_id  \
0  SAM'S CLUB          Calumet City          IL  600d813fadf77112c3bfa9a1   
1     SMITH'S                  Lehi          UT  60087e70c658dc12dc7354e1   
2     WALMART       West Palm Beach          FL  5fa6dbdcf54a65129aff01e9   
3         Y U  Feasterville Trevose          PA  5fbbe808fc5c9221743b89ce   
4      KROGER             Nashville          TN  5f38942e73dd9a14dbf9d914   

   transaction_date  total_sales  total_qty  barilla_sales  barilla_qty  
0  09/09/2021 11:04        79.51         12           6.54          1.0  
1  07/09/2021 16:08       151.97         51           1.59          1.0  
2  11/09/2021 09:52       257.16         86           1.28          1.0  
3  13/09/2021 11:38         1.89          1           1.89          1.0  
4  15/09/2021 18:07       274.64         86           1.59          1.0

In [52]:
#Select only the data for California
isCalifornia = df['store_state'] == 'CA'
df_cali = df[isCalifornia]
print("Number of stores in California: ", df_cali.shape[0])

df_cali.head()

Number of stores in California:  7469


store_name     store_city store_state                   user_id  \
5         WALMART          Dixon          CA  5ee356baa2881513e33f1a82   
13      FOOD4LESS  Moreno Valley          CA  5fbae3d1d7e73c26b279063b   
18         TARGET      San Ramon          CA  5e988812ad067313bf97dd6b   
32  SMART & FINAL         Clovis          CA  5f7cfc5c85dc3116b330ffde   
53      FOOD4LESS        Fontana          CA  5f1fa92bf1a92f14b284232b   

    transaction_date  total_sales  total_qty  barilla_sales  barilla_qty  \
5   19/09/2021 19:39       108.80         24           1.34          1.0   
13  16/09/2021 16:28        37.41         23           1.59          1.0   
18  06/09/2021 11:01       236.26         43           2.99          1.0   
32  09/09/2021 14:31        25.22         12           1.59          1.0   
53  17/09/2021 08:27       108.90         52           1.59          1.0   

    barilla_ms  
5         1.23  
13        4.25  
18        1.27  
32        6.30  
53        1.46

In [56]:
#Group the data by store_name and keep the first 25 by sales
df_cali_top = df_cali.groupby(by='store_name')[['total_sales','barilla_sales']].sum().sort_values(ascending=False, by='total_sales')[0:25]
df_cali_top.head()

total_sales  barilla_sales
store_name                             
WALMART        204532.09        4425.79
FOOD4LESS       89271.84        1224.25
WINCO FOODS     80388.80        1466.54
STATER BROS     67401.49        1696.38
SAFEWAY         50761.35        1901.61

In [57]:
#Add the custom column
df_cali_top['barilla_ms'] = round( df_cali_top['barilla_sales']/df_cali_top['total_sales'] * 100 ,2)
df_cali_top.head()

total_sales  barilla_sales  barilla_ms
store_name                                         
WALMART        204532.09        4425.79        2.16
FOOD4LESS       89271.84        1224.25        1.37
WINCO FOODS     80388.80        1466.54        1.82
STATER BROS     67401.49        1696.38        2.52
SAFEWAY         50761.35        1901.61        3.75

In [59]:
#Plot both the barilla_ms for each store in California
import plotly.express as px

px.bar(df_cali_top['barilla_ms'], labels={'value':'Total Sales', 'store_name':'Store Name'}).update_layout(showlegend=False)